### PR TITLE
rgw: s3select: add average function

### PR DESCRIPTION
**Pull request by albin-antony**
Signed-off-by: Albin Antony <aantony@redhat.com>

### CODE REVIEW

**Comment by albin-antony** [comment]
@galsalomon66 Please review

File: include/s3select_functions.h [hunk 1]
```diff
@@ -180,6 +181,37 @@ struct _fn_sum : public base_function
...
+        try
+        {
+            sum = sum + x->eval();
+            count = count + 1;
+        }
+        catch (base_s3select_exception &e)
+        {
+            std::cout << "illegal value for aggregation(sum). skipping." << std::endl;
+            if (e.severity() == base_s3select_exception::s3select_exp_en_t::FATAL)
+                throw;
```
Comment (galsalomon66) [inline include/s3select_functions.h:206]: use /base_s3select_exception/ to throw exception.
its also better to do the same fix for other functions (like sum)

Comment (albin-antony) [reply include/s3select_functions.h:206]: makes sense

File: include/s3select_functions.h [hunk 2]
```diff
@@ -26,7 +26,7 @@ struct push_4dig
 };
 static push_4dig g_push_4dig;
 
-enum class s3select_func_En_t {ADD,SUM,MIN,MAX,COUNT,TO_INT,TO_FLOAT,TO_TIMESTAMP,SUBSTR,EXTRACT,DATE_ADD,DATE_DIFF,UTCNOW};
+enum class s3select_func_En_t {ADD,SUM,AVG,MIN,MAX,COUNT,TO_INT,TO_FLOAT,TO_TIMESTAMP,SUBSTR,EXTRACT,DATE_ADD,DATE_DIFF,UTCNOW};
```
Comment (mattbenjamin) [inline include/s3select_functions.h:29]: please break across multiple lines:
enum class s3select_func_En_t {
ADD,
SUM,
...
(I think Gal's respacing commit will do this for all the code, but don't restore it)

File: include/s3select_functions.h [hunk 3]
```diff
@@ -180,6 +181,37 @@ struct _fn_sum : public base_function
...
+
+    bool operator()(std::vector<base_statement *> *args, variable *result)
+    {
+        std::vector<base_statement *>::iterator iter = args->begin();
+        base_statement *x = *iter;
+
+        try
+        {
+            sum = sum + x->eval();
+            count = count + 1;
```
Comment (mattbenjamin) [inline include/s3select_functions.h:200]: ok, works perfectly, but in c++ normally "count++")

File: include/s3select_functions.h (hunk 2 above)
Comment (albin-antony) [reply include/s3select_functions.h]: Modified Matt. Thanks for reviewing

File: include/s3select_functions.h [hunk 4]
```diff
@@ -180,6 +195,37 @@ struct _fn_sum : public base_function
...
+    bool operator()(std::vector<base_statement *> *args, variable *result)
+    {
+        std::vector<base_statement *>::iterator iter = args->begin();
+        base_statement *x = *iter;
+
+        try
+        {
+            sum = sum + x->eval();
+            count++;
+            throw;
```
Comment (galsalomon66) [inline include/s3select_functions.h:215]: rethrow , should be done on catch.

File: include/s3select_functions.h [hunk 5]
```diff
@@ -180,6 +195,37 @@ struct _fn_sum : public base_function
...
+        }
+        catch (base_s3select_exception &e)
+        {
+            if (e.severity() == base_s3select_exception::s3select_exp_en_t::FATAL)
```
Comment (galsalomon66) [inline include/s3select_functions.h:219]: its better to remove condition on severity, 
just re throw.

File: test/s3select_test.cpp [hunk 6]
```diff
@@ -238,3 +238,32 @@ TEST(TestS3SElect, utcnow_function)
...
+void generate_csv(std::string& out, size_t size) {
+  // schema is: int, float, string, string
+  std::stringstream ss;
+  for (auto i = 0U; i < size; ++i) {
+    ss << i << "," << i/10.0 << "," << "foo"+std::to_string(i) << "," << std::to_string(i)+"bar" << std::endl;
+  }
+  out = ss.str();
+}
+
+TEST(TestS3selectFunctions, avg)
```
Comment (galsalomon66) [inline test/s3select_test.cpp:250]: that's ok
its positive test, on constant values. (will stay with this)

generally, it need also for negative test (for testing the error flow), 
and also on random input.
future task.

File: include/s3select_functions.h (hunk 4 above)
Comment (albin-antony) [reply include/s3select_functions.h]: ack

**Comment by albin-antony** [comment]
> why using float for count?

Because, average result might be a floating point number. And that will be lost if we divide by an integer. So, dividing by a floating point number implicitly converts to a decimal number.

**Comment by mattbenjamin** [comment]
I didn't notice that--I'd declare count normally and divide by type(count)?  But also, I had noticed the use of float and wondered why float rather than double?  What is the general thinking around floating point results in the engine?

**Comment by albin-antony** [comment]
> I didn't notice that--I'd declare count normally and divide by type(count)? But also, I had noticed the use of float and wondered why float rather than double? What is the general thinking around floating point results in the engine?

Will use double. Thanks.

**Comment by albin-antony** [comment]
Modified @galsalomon66 @mattbenjamin

**Comment by galsalomon66** [comment]
> I didn't notice that--I'd declare count normally and divide by type(count)? But also, I had noticed the use of float and wondered why float rather than double? What is the general thinking around floating point results in the engine?

should use ``value`` object (it handle all s3-select types and operations) , 
what is probably missing is operator override for /

File: include/s3select_functions.h [hunk 7]
```diff
@@ -180,6 +193,35 @@ struct _fn_sum : public base_function
...
+        }
+        catch (base_s3select_exception &e)
+        {
+            throw base_s3select_exception(e.what());
+        }
+
+        return true;
+    }
+
+    virtual void get_aggregate_result(variable *result) { *result = sum/count ;}
```
Comment (galsalomon66) [inline include/s3select_functions.h:222]: what happens in the case count = 0? (empty stream, or wrong stream)

File: test/s3select_test.cpp [hunk 8]
```diff
@@ -238,3 +238,32 @@ TEST(TestS3SElect, utcnow_function)
...
+{
+    s3select s3select_syntax;
+    const std::string input_query = "select avg(int(_1)) from stdin;";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:253]: add float test.
add empty stream test.

**Pull request merged**